slsim/lens.py

```diff
@@ -517,69 +514,91 @@ def extended_source_magnitude(self, band, lensed=False):
-                light_model_list=kwargs_model.get("source_light_model_list", [])
-            )
-            lensModel = LensModel(
-                lens_model_list=kwargs_model.get("lens_model_list", [])
-            )
-            theta_E = self.einstein_radius
-            center_source = self.source.extended_source_position(
-                center_lens=self.deflector_position, draw_area=self.test_area
-            )
+            theta_E_list = self.einstein_radius
+            self._extended_source_magnification_list = []
+            # loop through each source.
+            for index, source in enumerate(self.source):
+                _light_model_list = kwargs_model.get(
+                        "source_light_model_list", [])[index]
+                kwargs_source_mag = [kwargs_params["kwargs_source"][index]]
+                if isinstance(_light_model_list, list):
+                    light_model_list = _light_model_list
+                else:
+                    light_model_list = [_light_model_list]
+                lightModel = LightModel(
+                    light_model_list=light_model_list)
+                lensModel = LensModel(
+                    lens_model_list=kwargs_model.get("lens_model_list", [])
+                )
+                theta_E = theta_E_list[index]
+                center_source = source.extended_source_position(
+                    center_lens=self.deflector_position, draw_area=self.test_area
+                )
 
-            kwargs_source_mag = kwargs_params["kwargs_source"]
-            kwargs_source_amp = data_util.magnitude2amplitude(
-                lightModel, kwargs_source_mag, magnitude_zero_point=0
-            )
+                kwargs_source_amp = data_util.magnitude2amplitude(
+                    lightModel, kwargs_source_mag, magnitude_zero_point=0
+                )
 
-            num_pix = 200
-            delta_pix = theta_E * 4 / num_pix
-            x, y = util.make_grid(numPix=num_pix, deltapix=delta_pix)
-            x += center_source[0]
-            y += center_source[1]
-            beta_x, beta_y = lensModel.ray_shooting(x, y, kwargs_params["kwargs_lens"])
-            flux_lensed = np.sum(
-                lightModel.surface_brightness(beta_x, beta_y, kwargs_source_amp)
-            )
-            flux_no_lens = np.sum(
-                lightModel.surface_brightness(x, y, kwargs_source_amp)
-            )
-            if flux_no_lens > 0:
-                self._extended_source_magnification = flux_lensed / flux_no_lens
-            else:
-                self._extended_source_magnification = 0
-        return self._extended_source_magnification
+                num_pix = 200
+                delta_pix = theta_E * 4 / num_pix
+                x, y = util.make_grid(numPix=num_pix, deltapix=delta_pix)
+                x += center_source[0]
+                y += center_source[1]
+                beta_x, beta_y = lensModel.ray_shooting(x, y, kwargs_params["kwargs_lens"])
```

Comment:
the lensModel() object needs to know the source redshift to do the ray-tracing, so that does not work. I need to make a new lenstronomy release (it's implemented but not on pip), then we can change the source redshift such that the ray_shooting() knows what source redshift to pick